gh-83714: Use statx on Linux 4.11 and later in os.stat

jbosboom opened the pull request:
I've posted some questions in gh-83714.  For now, this PR should be considered earnest effort (in the sense of [earnest money](https://en.wiktionary.org/wiki/earnest_money)).  It will change based on the answers to the questions in the issue.  It needs tests for the new features and the fallback that are not obvious to implement.  I need guidance as to when it should strictly conform to PEP 7 and when it should match the surrounding code style (especially with regards to brace placement).  But it should demonstrate I'm willing and able.

The libc statx wrapper is required at build time, but weakly linked.  There is no configure test; we just look for `STATX_BASIC_STATS` being defined by `sys/stat.h`.  `posix_do_stat` will fall back to calling stat if statx is not available at runtime due to an old libc or an old kernel.  The fallback variable `statx_works` is based on `dup3_works`, for the thread-safety of which see [@ericsnowcurrently's comment](https://github.com/python/cpython/pull/100262#issuecomment-1353516365).  The values are (to me) nonobvious: -1 means statx has never failed, 0 means its first failure was ENOSYS and 1 means its first failure was something other than ENOSYS.

This PR is not directly based on gh-19125, but was greatly informed by it, and I'd be happy to give Co-authored-by and/or blurb credit to @ntninja.

- Issue: gh-83714

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--136334.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

jbosboom committed:
bpo-39533: Use statx on Linux 4.11 and later in os.stat

The libc statx wrapper is required at build time, but weakly linked, and
posix_do_stat will fall back to calling stat if statx is not available
at runtime due to an old libc or an old kernel.

StanFromIreland commented on Doc/library/os.rst:
```diff
@@ -3276,6 +3279,34 @@ features:
+      See the :const:`!STATX_ATTR* <stat.STATX_ATTR_COMPRESSED>`
+      constants in the :mod:`stat` module.
+
+      .. versionadded: next
```
These need a double colon, otherwise they are parsed as comments.

jbosboom committed:
Fix versionadded documentation syntax

jbosboom committed:
Add blurb

jbosboom committed:
Fix typo stx_subvol -> st_subvol in docs

jbosboom committed:
Allow statx to be explicitly configured out for old glibc

jbosboom committed:
Implement os.statx

Co-authored-by: Erin of Yukis <erin-dev@ninetailed.ninja>

jbosboom committed:
Use correct function name in os.statx argument validation errors

cmaloney commented on Doc/library/os.rst:
```diff
@@ -3276,6 +3279,34 @@ features:
 
       User defined flags for file.
 
+   .. attribute:: st_attributes
```
These should probably get https://docs.python.org/3/library/intro.html#availability labels that they're only set on Linux

`.. availability:: Linux.`

For them may also make sense to use the manpage marker to direct people for full details: 
```
:manpage:`statx(2)`
```

cmaloney commented on Doc/library/os.rst:
```diff
@@ -3367,6 +3398,11 @@ features:
       Added the :attr:`st_birthtime` member on Windows.
 
+   .. versionchanged:: next
+      Added the :attr:`st_birthtime`, :attr:`st_attributes`,
```
Should `st_birthtime_ns` also be included/set here?

cmaloney commented on Doc/library/os.rst:
```diff
@@ -3276,6 +3279,34 @@ features:
+
+   .. attribute:: st_attributes_mask
+
+      Linux file attributes supported by the filesystem containing the file.
```
It would be nice nice to link to the enum / constants available here section

cmaloney commented on Doc/library/stat.rst:
```diff
@@ -493,3 +493,22 @@ constants, but are not an exhaustive list.
+On Linux, the following constants are available for comparing against the
+``st_attributes`` and ``st_attributes_mask`` members returned by
+:func:`os.stat`.  See the `statx(2) man page
+<https://man.archlinux.org/man/statx.2#File_attributes>` for more detail on the
```
I think this can / should use:
```
:manpage:`statx(2)`
```

jbosboom replied:
Done, though this generates a link to the Debian man pages, so some of these attributes are missing in the man page.

cmaloney commented on Modules/posixmodule.c:
```diff
@@ -408,6 +408,30 @@ extern char *ctermid_r(char *);
+#ifdef HAVE_LINUX_STATX
+#  pragma weak statx
+/* provide constants introduced later than statx itself */
+#  ifndef STATX_MNT_ID
```
Rather than hardcode, can we get these values from / rely on `<sys/stat.h>`

jbosboom replied:
No.  statx is extended in new kernel releases by giving reserved space in `struct statx` proper names and adding new mask bits for them, so the member names and mask constants available at build time vary based on the version of kernel API headers installed on the build system.  I would not want to document that a field is only available on Python 3.15 _built on a host with kernel API headers >= 5.x_, especially because some build environments specifically use old versions of everything in an attempt to minimize dependencies.

cmaloney reviewed (commented):
Some docs suggestions / CPython sphinx has special tags for a couple of the things; haven't looked at main implementation a lot yet

gpshead commented on aclocal.m4:
Was this aclocal.m4 update required by this PR?

jbosboom replied:
I don't know.  This is the first time I've had to modify anything autoconf-related.  This was among the files updated when I ran autoreconf, and the file is version-controlled, so I committed it.

gpshead replied:
Try https://devguide.python.org/getting-started/setup-building/#regenerate-configure which should regenerate configure with the same version we use in the repo (it should not update aclocal.m4)

gpshead commented on configure.ac:
```diff
@@ -7618,6 +7618,22 @@ AC_ARG_ENABLE([test-modules],
+    [AC_DEFINE(HAVE_LINUX_STATX, 1, Define if you have the 'statx' function.)
+     AC_MSG_RESULT(yes)],
+    [AC_MSG_RESULT(no)
+     AC_MSG_ERROR([statx syscall wrapper function not available; upgrade to
```
Wouldn't it be simpler for this not to be an error and yet another --with flag, but just to have the undef HAVE_LINUX_STATX state remain as it does on non-Linux platforms?  automatic detection is all anyone actually wants.  One less knob that won't be used on our configure script.

jbosboom replied:
I interpreted [your comment on Discourse](https://discuss.python.org/t/minimum-supported-linux-kernel-and-glibc-versions/103429/2) to mean 1) that the default should be enabled 2) but configure should check 3) and there should be an opt-out flag for builds targeting old Android.  I interpret this GitHub comment to mean that the default should be disabled, with the check enabling, and no opt-out flag for builds targeting old Android.

I don't feel strongly either way.  Just be clear what you'd like. :)  (See also [my recent comment in the issue](https://github.com/python/cpython/issues/83714#issuecomment-3252649070).)

gpshead replied:
my goal is lower complexity in the configure.ac.  if nothing in our PR CI or buildbots needs configure flag, we're fine with just the check.

jbosboom replied:
The Android builds succeeded with the standard automatic detection macro.

gpshead commented on Modules/posixmodule.c:
```diff
@@ -2779,16 +2838,152 @@ _pystat_fromstructstat(PyObject *module, STRUCT_STAT *st)
 #ifdef HAVE_STRUCT_STAT_ST_REPARSE_TAG
     SET_ITEM(ST_REPARSE_TAG_IDX, PyLong_FromUnsignedLong(st->st_reparse_tag));
 #endif
+#ifdef HAVE_LINUX_STATX
```
I don't believe `os.statx` should return an `os.stat_result`.  This complicates the existing PySequenceStruct making it larger and slower to create for what are going to most often by empty values on all platforms and the most common `os.stat` calls.

Keep the birthtime code above as we're now exposing that on one more platform for more consistency from os.stat itself.

But os.statx can have a nice new namespace style return type using its class for default field values instead of allocating space for them no matter what in every instance regardless of if they've been filled in.

jbosboom committed:
Merge branch 'main' into statx

jbosboom committed:
Move statx-specific members from os.stat_result to new os.statx_result

jbosboom committed:
make regen-configure

jbosboom committed:
Reimplement os.statx_result using PyMemberDef and PyGetSetDef; remove raw kwarg

We no longer return None for invalid attributes for the benefit of
drop-in replacement of os.stat when not all basic stats are required or
sync=False is acceptable.  Such code was already accepting faked values
from os.stat, so can also accept faked values from os.statx.  Remove the
raw kwarg as it would have no effect.

jbosboom committed:
Remove statx-specific members from os.stat_result documentation

These members were removed from os.stat_result in a previous commit.

jbosboom committed:
Simplify statx configure check

jbosboom committed:
Document os.statx

jbosboom committed:
Prepare for review

jbosboom committed:
Document st_birthtime fallback behavior

cmaloney commented on Doc/library/os.rst:
```diff
@@ -3364,6 +3370,139 @@ features:
+
+.. function:: statx(path, mask, *, dir_fd=None, follow_symlinks=True, sync=None)
+
+   Get the status of a file or file descriptor by performing a :c:func:`statx`
```
Should be able to make the missing c func / struct pieces stop erroring by adding to: https://github.com/python/cpython/blob/4e00e2504fdcf06b5d55a773b31f49b1d3d6d35c/Doc/conf.py#L121 (in general try with new things to not add new warnings; are people working on resolving existing sphinx warnings)

StanFromIreland replied:
```suggestion
   Get the status of a file or file descriptor by performing a :c:func:`!statx`
```
This will also work.

jbosboom replied:
Running `make html` and `make check` as [directed by the devguide](https://devguide.python.org/documentation/start-documenting/), I don't see any warnings.

cmaloney replied:
Getting all the warnings out of the docs build is difficult (there are many warnings in non-changed code which don't matter for PRs; but new ones do); the CI does check for new warnings and attaches them to PRs. I can see some here on the Github UI:

<img width="779" height="319" alt="image" src="https://github.com/user-attachments/assets/1e7476c5-f209-4569-915d-3caa5b0fa676" />

cmaloney commented on Doc/library/os.rst:
```diff
@@ -3364,6 +3370,139 @@ features:
+   actually retrieved (which may differ from *mask*).
+
+   The optional parameter *sync* controls the freshness of the returned
+   information.  ``sync=True`` requests that the kernel return up-to-date
```
Should document the value of `None` which is the default; may make sense to just make it always True/False for now

cmaloney commented on Lib/test/test_os.py:
```diff
@@ -660,21 +668,15 @@ def trunc(x): return x
-            self.assertAlmostEqual(floaty, nanosecondy, delta=2)
-
-        # Ensure both birthtime and birthtime_ns roughly agree, if present
+        time_attributes = 'st_atime st_mtime st_ctime'.split()
```
nit: why not just make a three element list here? (I see the old code did this, but seems reasonable to just update)

cmaloney commented on Lib/test/test_os.py:
```diff
@@ -735,6 +737,85 @@ def test_stat_result_pickle(self):
+    def check_statx_attributes(self, fname):
+        maximal_mask = 0
+        for name in dir(os):
+            if name.startswith('STATX_'):
```
feels like there should be a better way to group / gather these (maybe have them all live as constants in a Python class that acts as a namespace?)

jbosboom replied:
Similar constants for other `os` functions are also dumped in the module namespace.  That doesn't mean we can't do something different, but we'd be setting somewhat of a precedent.

(It would be nice to have an `IntFlag` enum class that can report which members are valid given a bitmask, but adding an import dependency to `os` doesn't seem like a good idea.)

cmaloney replied:
I'm fairly strongly against "find everything that starts with `STATX_` in os" as that will just get slower as `os` gets larger. 

Can be just a manually maintained list for now; would be nice to do something more namespace like (but also don't want to add a new CPython standard here).

cmaloney commented on Lib/test/test_os.py:
```diff
@@ -735,6 +737,85 @@ def test_stat_result_pickle(self):
+        for sync in (False, True):
+            with self.subTest(sync=sync):
+                os.statx(self.fname, os.STATX_BASIC_STATS, sync=sync)
+
```
Will need more test cases for different ways of constructing/using statx (path_fd, dir_fd, PathLike)

jbosboom replied:
Should the dir_fd tests go in here in test_os.py or in TestPosixDirFd in test_posix.py? 

Should all the statx tests be in test_posix.py?  The comment at the top of test_os.py says it's for "a few functions which have been determined to be more portable than they had been thought to be", and statx is not.

cmaloney replied:
I think in `test_os` works; don't need to test that "do the dir_fd helpers use handle every case right", want to just make sure "if someone deleted the dir_fd handling code; a test would fail"

cmaloney replied:
`test_posix` seems to be around https://docs.python.org/3/library/posix.html; I don't have experience / knowledge around that so not sure

cmaloney commented on Modules/posixmodule.c:
```diff
@@ -2829,6 +2938,38 @@ posix_do_stat(PyObject *module, const char *function_name, path_t *path,
+        int flags = AT_NO_AUTOMOUNT;
+        flags |= follow_symlinks ? 0 : AT_SYMLINK_NOFOLLOW;
+
+        Py_BEGIN_ALLOW_THREADS
```
nit: I don't see a reason not to start allowing threads a bit earlier; aren't touching the interpreter in the lines just before this

cmaloney commented on Modules/posixmodule.c:
```diff
@@ -3277,6 +3418,322 @@ os_lstat_impl(PyObject *module, path_t *path, int dir_fd)
+typedef struct {
+    PyObject_HEAD
+    struct statx stx;
+#if STATX_RESULT_CACHE_SLOTS > 0
```
isn't this always > 0? (the constant 17)

jbosboom replied:
It's defined this way to allow reviewers and future maintainers to more easily experiment with not having a cache (because member arrays can't be empty in C).  On reflection, it doesn't help much, because the no-cache configuration would change most of the getset defs to members for faster access, and after that, commenting out the member is trivial.

cmaloney commented on Modules/posixmodule.c:
```diff
@@ -3277,6 +3418,322 @@ os_lstat_impl(PyObject *module, path_t *path, int dir_fd)
+    int result;
+    /* Future bits may refer to members beyond the current size of struct
+       statx, so we need to mask them off to prevent memory corruption. */
+    mask &= _Py_STATX_KNOWN;
```
Should this error if the mask was changed by this code?

jbosboom replied:
Hmm, maybe.  We'd only adjust the mask for Python code that specified the mask numerically instead of using the `os.STATX_*` constants.  That's probably only code that wants to use members Python 3.16 knows about but Python 3.15 doesn't.  `os.statx_result` doesn't provide Python with an interface to access members the interpreter doesn't know about, so while code could use the numeric value of `STATX_FOO`, accessing `stx_foo` would raise `AttributeError`.  And that's how current code copes with `os.stat_result`'s members varying across versions.  But I'm open to arguments that failing fast is better and version-crossing code should instead change their mask based on the constants available.

cmaloney commented on Modules/posixmodule.c:
```diff
@@ -17878,6 +18336,30 @@ all_ins(PyObject *m)
 #endif  /* HAVE_EVENTFD && EFD_CLOEXEC */
 
+#ifdef HAVE_STATX
+    if (PyModule_AddIntMacro(m, STATX_TYPE)) return -1;
```
nit: Any reason not to alphabetize this list? 

I like having either a comment why not alphabetical or to alphabetize for easier later code reading "is this member in the list" when trying to add more

jbosboom replied:
The order matches the definitions in `/usr/include/linux/stat.h`.  When new bits are defined, they're added at the end (there are no gaps in the definitions).  Alphabetizing would make things harder, not easier.

cmaloney reviewed (commented):
Reviewed the pieces I have reasonable knowledge of; don't have enough experience with macros and defining new structs to go in more detail through those pieces. For first contributions to CPython really impressive in getting things to the CPython style (docs, C code style, etc)

High level:
1. This PR is large for CPython; Smaller chunks will make it easier to review and land; My thought would be adding one part adding os.statx; second updating os.stat to use statx when available
2. The `os.stat` change will need (on newer + older kernels). Just adding statx doesn't need as much performance validation
    1. [`pyperf`](https://pyperf.readthedocs.io/en/latest/) microbenchmark that shows stat didn't get slower
    1. Validate that Python startup time does not change significantly (I believe github.com/python/pyperformance has two benchmarks around that)
3. `statx` will probably make sense to add to `pathlib.Path` for easy access like `stat` is (but that should probably be another PR with the same issue)

With the thoroughness you're doing definitely feels like will get to the confidence to change `os.stat` to point to `os.statx`. Again; really impressive for a first CPython contributions.

jbosboom committed:
Implement os.stat with statx only if the statx syscall works

glibc doesn't remember that statx previously failed with ENOSYS, so on
old kernels, glibc's emulation of statx via stat has the overhead of a
failed syscall.  Only call the statx wrapper under the same conditions
that we make os.statx available.

statx_works was always a global C variable, but now its scope is the
file instead of just posix_do_stat.  Whether statx works is a
process-global property, so using a global C variable is appropriate.
Previously access to statx_works was unsynchronized, following the
pattern of dup3_works.  Now we rely on posixmodule_exec happening-before
any thread can call os.stat.

jbosboom committed:
Test statx PathLike and dir_fd handling

Also two minor fixes to statx tests in test_os.py.

jbosboom committed:
statx docs fixes
